applications/serial_lte_modem/src/slm_util.c

```diff
@@ -7,65 +7,129 @@
 #include <string.h>
 #include <errno.h>
 #include <stdio.h>
+#include <nrf_errno.h>
 #include "slm_util.h"
 #include "slm_at_host.h"
 
-LOG_LEVEL_SET(CONFIG_SLM_LOG_LEVEL);
+LOG_MODULE_REGISTER(slm_util, CONFIG_SLM_LOG_LEVEL);
 
-/**
- * @brief Compare string ignoring case
- */
-bool slm_util_casecmp(const char *str1, const char *str2)
+static bool cmd_has_lower(const char *cmd)
 {
-	int str2_len = strlen(str2);
+	/* Set/test command names are delimited by '=', read by '?'. */
+	for (size_t i = 0; cmd[i] != '\0'; ++i) {
+		const char c = cmd[i];
+
+		if (c == '=' || c == '?') {
+			break;
+		} else if (islower(c)) {
+			LOG_ERR("FIX ME: AT command \"%s\" must be all-uppercase.", cmd);
+			return true;
+		}
+	}
+	return false;
+}
 
-	if (strlen(str1) != str2_len) {
-		return false;
+int slm_util_at_printf(const char *fmt, ...)
+{
+	int ret;
+	char buf[128];
+	va_list args;
+
+	va_start(args, fmt);
+	ret = vsnprintf(buf, sizeof(buf), fmt, args);
+	va_end(args);
+	if (ret >= sizeof(buf)) {
+		LOG_ERR("AT command \"%.16s...\" would get truncated from %u to %u bytes. "
+			"The buffer needs to be made bigger.", buf, ret, sizeof(buf) - 1);
+		return -E2BIG;
 	}
 
-	for (int i = 0; i < str2_len; i++) {
-		if (toupper((int)*(str1 + i)) != toupper((int)*(str2 + i))) {
-			return false;
-		}
+	/* AT command interception is case-sensitive. */
+	if (cmd_has_lower(buf)) {
+		return -EINVAL;
 	}
 
-	return true;
+	/* We are not interested in the AT response, so avoid having it copied.
+	 * However the modem library requires "OK\r\n" in the buffer to accept
+	 * a successful response if the command is intercepted so make room for it (5 bytes).
+	 */
+	ret = nrf_modem_at_cmd(buf, 5, "%s", buf);
+	if (ret == -NRF_E2BIG) {
+		ret = 0;
+	}
+	return ret;
 }
 
+int slm_util_at_scanf(const char *cmd, const char *fmt, ...)
+{
+	char buf[128];
+	va_list args;
+	int ret;
 
-/**
- * @brief Compare name of AT command ignoring case
- */
-bool slm_util_cmd_casecmp(const char *cmd, const char *slm_cmd)
+	/* AT command interception is case-sensitive. */
+	if (cmd_has_lower(cmd)) {
+		return -EINVAL;
+	}
+
+	ret = nrf_modem_at_cmd(buf, sizeof(buf), "%s", cmd);
+	if (ret == -NRF_E2BIG) {
+		LOG_ERR("AT response to \"%s\" truncated to %u bytes. "
+			"The buffer needs to be made bigger.", cmd, sizeof(buf) - 1);
+		buf[sizeof(buf) - 1] = '\0';
+	} else if (ret < 0) {
+		return ret;
+	}
+
+	va_start(args, fmt);
+	ret = vsscanf(buf, fmt, args);
+	va_end(args);
+
+	if (ret == 0) {
+		ret = -NRF_EBADMSG;
+	}
+	return ret;
+}
+
+int slm_util_at_cmd_fwd_from_cb(char *buf, size_t len, char *at_cmd)
+{
+/* Purposefully bypass nrf_modem_at_cmd() to avoid infinite recursion. */
```

Comment:
Comment meaning is bit obscure. I take it that this callback is to be called from nrf_modem_at_cmd and that is why we are redefining the nrf_modem_at_printf, to be able to print it without infinite recursion.